File: sop/engine.py
"""SOP 执行引擎 — 阶段门禁检查与流转控制。

SOPExecutor 是 FlowForge 与 LoopExecutor 之间的桥接层：
- SOPExecutor 管控阶段门禁（hard_rules / pitfalls）
- LoopExecutor / WorkflowExecutor 管控阶段内的实际任务执行

执行模型：
1. SOPExecutor 在每个阶段开始前检查所有 hard_rules
2. 若有 blocker 未通过则停止，等待修复后重试
3. 若所有 blocker 通过，则交由 LoopExecutor 执行阶段内任务
4. 阶段完成后记录结果，自动推进到下一阶段
5. 持续直到所有阶段完成或被 blocker 阻断

关键设计（遵守铁律）：
- 不直接执行任务，只做门禁检查和阶段推进（铁律3：组合优于继承）
- 不直接操作数据库（铁律4），状态由 CheckpointManager 持久化
- 不硬编码路径/密钥（铁律5），所有配置从 YAML 加载
"""
from __future__ import annotations

from datetime import datetime
from pathlib import Path
from typing import Any

import yaml

from flowforge.core.tracing import get_logger
from flowforge.sop.models import (
    HardRule,
    Pitfall,
    PredicateResult,
    Severity,
    SOPDefinition,
    SOPExecutionResult,
    SOPExecutionState,
    SOPStage,
    SOPStageResult,
)
from flowforge.sop.predicate import PredicateChecker

logger = get_logger("sop.engine")


class SOPExecutor:
    """SOP 执行器 — 阶段门禁检查与流转控制。

    SOPExecutor 不直接执行任务，只负责：
    1. 检查每个阶段的 hard_rules（blocker 未通过则阻断）
    2. 收集 pitfalls 警告
    3. 推进或回退阶段
    4. 记录执行结果

    实际任务执行由 LoopExecutor / WorkflowExecutor 完成，SOPExecutor
    通过 suggested_skill 字段提示应使用哪个 skill。

    Usage:
        checker = PredicateChecker()
        sop_def = load_sop_from_yaml("flowforge/config/sops/development.yaml")
        executor = SOPExecutor(sop_def, checker)
        result = await executor.execute_sop(feature_id="feat-001", context={...})
    """

    def __init__(
        self,
        sop_definition: SOPDefinition,
        predicate_checker: PredicateChecker,
    ) -> None:
        """初始化 SOP 执行器。

        Args:
            sop_definition: SOP 定义（包含所有阶段）
            predicate_checker: 谓词检查器实例
        """
        self._sop = sop_definition
        self._checker = predicate_checker
        self._state = SOPExecutionState(sop_id=sop_definition.id)
        logger.info(
            f"SOPExecutor: initialized with SOP '{sop_definition.id}' "
            f"({len(sop_definition.stages)} stages)"
        )

    @property
    def sop(self) -> SOPDefinition:
        """返回当前 SOP 定义。"""
        return self._sop

    @property
    def state(self) -> SOPExecutionState:
        """返回当前执行状态。"""
        return self._state

    def get_current_stage(self) -> SOPStage:
        """获取当前阶段。

        Returns:
            当前 SOPStage

        Raises:
            IndexError: 所有阶段已完成
        """
        if self._state.stage_index >= len(self._sop.stages):
            raise IndexError(
                f"stage_index {self._state.stage_index} out of range "
                f"(total {len(self._sop.stages)} stages)"
            )
        return self._sop.stages[self._state.stage_index]

    def get_stage(self, stage_id: str) -> SOPStage | None:
        """按 ID 查找阶段。"""
        for stage in self._sop.stages:
            if stage.id == stage_id:
                return stage
        return None

    def advance_stage(self) -> bool:
        """推进到下一阶段。

        Returns:
            True 表示成功推进，False 表示已到达最后一个阶段
        """
        if self._state.stage_index >= len(self._sop.stages) - 1:
            self._state.completed = True
            logger.info("SOPExecutor: reached final stage, SOP completed")
            return False
        self._state.stage_index += 1
        next_stage = self._sop.stages[self._state.stage_index]
        logger.info(
            f"SOPExecutor: advanced to stage {self._state.stage_index} "
            f"'{next_stage.id}' ({next_stage.label})"
        )
        return True

    def get_progress(self) -> dict[str, Any]:
        """获取执行进度。

        Returns:
            包含 current_stage_index / total_stages / completed_stages /
            remaining_stages / current_stage_id / is_completed 的字典
        """
        total = len(self._sop.stages)
        current = self._state.stage_index
        return {
            "sop_id": self._sop.id,
            "current_stage_index": current,
            "total_stages": total,
            "completed_stages": current,
            "remaining_stages": max(0, total - current),
            "current_stage_id": (
                self._sop.stages[current].id if current < total else ""
            ),
            "current_stage_label": (
                self._sop.stages[current].label if current < total else ""
            ),
            "is_completed": self._state.completed,
        }

    async def _check_rule(
        self,
        rule: HardRule | Pitfall,
        context: dict[str, Any],
    ) -> tuple[PredicateResult, dict[str, Any]]:
        """检查单条规则（hard_rule 或 pitfall）。

        Returns:
            (PredicateResult, 规则摘要 dict) 二元组
        """
        result = await self._checker.check(rule.predicate, context=context)
        summary = {
            "rule_id": rule.id,
            "text": rule.text,
            "severity": rule.severity.value,
            "passed": result.passed,
            "message": result.message,
            "evidence": result.evidence,
        }
        return result, summary

    async def execute_stage(
        self,
        stage_id: str,
        context: dict[str, Any] | None = None,
    ) -> SOPStageResult:
        """执行单个阶段的门禁检查。

        流程：
        1. 查找指定阶段
        2. 检查所有 hard_rules（blocker 未通过则阻断）
        3. 检查所有 pitfalls（收集警告）
        4. 返回阶段结果

        Args:
            stage_id: 阶段标识
            context: 运行时上下文（可包含 last_command / command_history /
                     author / reviewer / guardian / feature_doc 等）

        Returns:
            SOPStageResult 阶段执行结果
        """
        ctx = context or {}
        stage = self.get_stage(stage_id)
        if stage is None:
            return SOPStageResult(
                stage_id=stage_id,
                stage_label="",
                passed=False,
                blocker_messages=[f"stage '{stage_id}' not found in SOP '{self._sop.id}'"],
                executed_at=datetime.utcnow(),
            )

        logger.info(
            f"SOPExecutor: executing stage '{stage.id}' ({stage.label}) "
            f"with {len(stage.hard_rules)} hard_rules, {len(stage.pitfalls)} pitfalls"
        )

        hard_rule_results: list[dict[str, Any]] = []
        pitfall_results: list[dict[str, Any]] = []
        blocker_messages: list[str] = []
        warning_messages: list[str] = []

        # 检查 hard_rules
        for rule in stage.hard_rules:
            result, summary = await self._check_rule(rule, ctx)
            hard_rule_results.append(summary)
            if not result.passed:
                msg = f"[{rule.id}] {rule.text} — {result.message}"
                if rule.severity == Severity.BLOCKER:
                    blocker_messages.append(msg)
                    logger.warning(
                        f"SOPExecutor: stage '{stage.id}' hard_rule '{rule.id}' BLOCKED: {result.message}"
                    )
                else:
                    warning_messages.append(msg)
                    logger.info(
                        f"SOPExecutor: stage '{stage.id}' hard_rule '{rule.id}' warning: {result.message}"
                    )

        # 检查 pitfalls
        for pitfall in stage.pitfalls:
            result, summary = await self._check_rule(pitfall, ctx)
            pitfall_results.append(summary)
            if not result.passed:
                msg = f"[{pitfall.id}] {pitfall.text} — {result.message}"
                if pitfall.severity == Severity.BLOCKER:
                    # pitfall 的 blocker 仅在非可选阶段才阻断
                    if stage.optional:
                        warning_messages.append(
                            f"(optional stage) {msg}"
                        )
                    else:
                        blocker_messages.append(msg)
                        logger.warning(
                            f"SOPExecutor: stage '{stage.id}' pitfall '{pitfall.id}' BLOCKED: {result.message}"
                        )
                else:
                    warning_messages.append(msg)
                    logger.info(
                        f"SOPExecutor: stage '{stage.id}' pitfall '{pitfall.id}' warning: {result.message}"
                    )

        # 可选阶段失败不阻断主流程
        stage_passed = (
            len(blocker_messages) == 0 if not stage.optional else True
        )
        if stage.optional and blocker_messages:
            # 可选阶段的 blocker 降级为 warning
            warning_messages.extend(
                f"(optional, downgraded) {m}" for m in blocker_messages
            )
            blocker_messages.clear()
            stage_passed = True
            logger.info(
                f"SOPExecutor: optional stage '{stage.id}' blockers downgraded to warnings"
            )

        result = SOPStageResult(
            stage_id=stage.id,
            stage_label=stage.label,
            passed=stage_passed,
            hard_rule_results=hard_rule_results,
            pitfall_results=pitfall_results,
            blocker_messages=blocker_messages,
            warning_messages=warning_messages,
            executed_at=datetime.utcnow(),
        )

        # 记录到执行状态
        self._state.stage_results[stage.id] = result.model_dump()

        logger.info(
            f"SOPExecutor: stage '{stage.id}' result: "
            f"passed={stage_passed}, blockers={len(blocker_messages)}, warnings={len(warning_messages)}"
        )
        return result

    async def execute_sop(
        self,
        feature_id: str,
        context: dict[str, Any] | None = None,
    ) -> SOPExecutionResult:
        """执行完整 SOP — 按顺序执行所有阶段。

        流程：
        1. 初始化执行状态
        2. 按顺序执行每个阶段
        3. 阶段通过后自动推进到下一阶段
        4. 遇到 blocker 则停止并返回部分结果
        5. 所有阶段完成后标记 success=True

        Args:
            feature_id: feature 标识
            context: 运行时上下文

        Returns:
            SOPExecutionResult 完整执行结果
        """
        ctx = context or {}
        started_at = datetime.utcnow()
        self._state = SOPExecutionState(
            sop_id=self._sop.id,
            feature_id=feature_id,
            stage_index=0,
            started_at=started_at,
            completed=False,
        )

        logger.info(
            f"SOPExecutor: starting SOP '{self._sop.id}' for feature '{feature_id}' "
            f"with {len(self._sop.stages)} stages"
        )

        stage_results: list[SOPStageResult] = []
        all_blocker_messages: list[str] = []
        all_warning_messages: list[str] = []
        final_stage_id = ""
        success = False

        for index, stage in enumerate(self._sop.stages):
            self._state.stage_index = index
            final_stage_id = stage.id

            stage_result = await self.execute_stage(stage.id, ctx)
            stage_results.append(stage_result)
            all_blocker_messages.extend(stage_result.blocker_messages)
            all_warning_messages.extend(stage_result.warning_messages)

            if not stage_result.passed:
                logger.warning(
                    f"SOPExecutor: SOP halted at stage '{stage.id}' due to blockers"
                )
                break

            logger.info(
                f"SOPExecutor: stage '{stage.id}' passed, advancing"
            )
        else:
            # 所有阶段都通过（for-else：未 break）
            success = True
            self._state.completed = True
            logger.info(
                f"SOPExecutor: SOP '{self._sop.id}' completed successfully for feature '{feature_id}'"
            )

        completed_at = datetime.utcnow()
        return SOPExecutionResult(
            sop_id=self._sop.id,
            feature_id=feature_id,
            success=success,
            stage_results=stage_results,
            final_stage_id=final_stage_id,
            blocker_messages=all_blocker_messages,
            warning_messages=all_warning_messages,
            started_at=started_at,
            completed_at=completed_at,
        )

    def reset(self) -> None:
        """重置执行状态到初始值。

        用于重新执行 SOP 或在 blocker 修复后从头开始。
        """
        logger.info(f"SOPExecutor: resetting state for SOP '{self._sop.id}'")
        self._state = SOPExecutionState(sop_id=self._sop.id)

    def resume_from_current(self) -> None:
        """从当前阶段继续执行（不重置 stage_index）。

        用于在 blocker 修复后从断点处继续。
        """
        logger.info(
            f"SOPExecutor: resuming from stage_index={self._state.stage_index}"
        )
        self._state.completed = False


def load_sop_from_yaml(yaml_path: str | Path) -> SOPDefinition:
    """从 YAML 文件加载 SOP 定义。

    Args:
        yaml_path: YAML 文件路径

    Returns:
        SOPDefinition 实例
    """
    path = Path(yaml_path)
    logger.info(f"load_sop_from_yaml: loading SOP from {path}")
    with open(path, encoding="utf-8") as f:
        data = yaml.safe_load(f)
    return SOPDefinition.model_validate(data)


def load_sops_from_dir(sops_dir: str | Path) -> dict[str, SOPDefinition]:
    """从目录加载所有 SOP 定义。

    Args:
        sops_dir: 包含 SOP YAML 文件的目录

    Returns:
        {sop_id: SOPDefinition} 字典
    """
    dir_path = Path(sops_dir)
    sops: dict[str, SOPDefinition] = {}
    if not dir_path.exists():
        logger.warning(f"load_sops_from_dir: directory {dir_path} does not exist")
        return sops

    for yaml_file in dir_path.glob("*.yaml"):
        try:
            sop = load_sop_from_yaml(yaml_file)
            sops[sop.id] = sop
            logger.info(f"load_sops_from_dir: loaded SOP '{sop.id}' from {yaml_file.name}")
        except Exception as exc:
            logger.error(f"load_sops_from_dir: failed to load {yaml_file.name}: {exc}")
    return sops

File: packages/forgekin/sop/src/engine.ts
/**
 * @flowforge/forgekin-sop — SOP 执行引擎
 *
 * 对齐 Python `sop/engine.py`：阶段门禁检查与流转控制。
 * SOPExecutor 是 FlowForge 与 LoopExecutor 之间的桥接层：
 * - SOPExecutor 管控阶段门禁（hardRules / pitfalls）
 * - LoopExecutor / WorkflowExecutor 管控阶段内的实际任务执行
 *
 * 关键设计（遵守铁律）：
 * - 不直接执行任务，只做门禁检查和阶段推进（铁律3）
 * - 不直接操作数据库（铁律4），状态由 CheckpointManager 持久化
 * - 不硬编码路径/密钥（铁律5），所有配置从 YAML 加载
 */
import { promises as fs } from 'node:fs';
import { basename, resolve } from 'node:path';
import { parse as parseYaml } from 'yaml';
import {
  HardRule,
  makePredicateConfig,
  makeSOPDefinition,
  makeSOPExecutionState,
  makeSOPStage,
  Pitfall,
  PredicateConfig,
  PredicateContext,
  PredicateResult,
  RuleSummary,
  Severity,
  SOPDefinition,
  SOPExecutionResult,
  SOPExecutionState,
  SOPStage,
  SOPStageResult,
} from './models.js';
import { PredicateChecker } from './predicate.js';

/** 执行进度快照（对齐 get_progress 返回 dict，字段转 camelCase） */
export interface SOPProgress {
  readonly sopId: string;
  readonly currentStageIndex: number;
  readonly totalStages: number;
  readonly completedStages: number;
  readonly remainingStages: number;
  readonly currentStageId: string;
  readonly currentStageLabel: string;
  readonly isCompleted: boolean;
}

/**
 * SOP 执行器 — 阶段门禁检查与流转控制（对齐 SOPExecutor）。
 *
 * 只负责：
 * 1. 检查每个阶段的 hard_rules（blocker 未通过则阻断）
 * 2. 收集 pitfalls 警告
 * 3. 推进或回退阶段
 * 4. 记录执行结果
 *
 * 实际任务执行由 LoopExecutor / WorkflowExecutor 完成，SOPExecutor
 * 通过 suggestedSkill 字段提示应使用哪个 skill。
 */
export class SOPExecutor {
  /** 当前 SOP 定义（对齐 sop 属性） */
  readonly sop: SOPDefinition;
  readonly checker: PredicateChecker;
  /** 当前执行状态（对齐 state 属性） */
  state: SOPExecutionState;

  constructor(sopDefinition: SOPDefinition, predicateChecker: PredicateChecker) {
    this.sop = sopDefinition;
    this.checker = predicateChecker;
    this.state = makeSOPExecutionState({ sopId: sopDefinition.id });
  }

  /** 获取当前阶段（所有阶段已完成时抛 RangeError，对齐 IndexError） */
  getCurrentStage(): SOPStage {
    if (this.state.stageIndex >= this.sop.stages.length) {
      throw new RangeError(
        `stage_index ${this.state.stageIndex} out of range (total ${this.sop.stages.length} stages)`,
      );
    }
    return this.sop.stages[this.state.stageIndex]!;
  }

  /** 按 ID 查找阶段（未找到返回 null） */
  getStage(stageId: string): SOPStage | null {
    return this.sop.stages.find((stage) => stage.id === stageId) ?? null;
  }

  /**
   * 推进到下一阶段（对齐 advance_stage）。
   *
   * @returns true 表示成功推进；false 表示已到达最后一个阶段（标记 completed）
   */
  advanceStage(): boolean {
    if (this.state.stageIndex >= this.sop.stages.length - 1) {
      this.state.completed = true;
      return false;
    }
    this.state.stageIndex += 1;
    return true;
  }

  /** 获取执行进度（对齐 get_progress） */
  getProgress(): SOPProgress {
    const total = this.sop.stages.length;
    const current = this.state.stageIndex;
    const currentStage = current < total ? this.sop.stages[current] : undefined;
    return {
      sopId: this.sop.id,
      currentStageIndex: current,
      totalStages: total,
      completedStages: current,
      remainingStages: Math.max(0, total - current),
      currentStageId: currentStage?.id ?? '',
      currentStageLabel: currentStage?.label ?? '',
      isCompleted: this.state.completed,
    };
  }

  /** 检查单条规则（对齐 _check_rule），返回 (result, summary) */
  async checkRule(
    rule: HardRule | Pitfall,
    context: PredicateContext,
  ): Promise<{ result: PredicateResult; summary: RuleSummary }> {
    const result = await this.checker.check(rule.predicate, context);
    const summary: RuleSummary = {
      ruleId: rule.id,
      text: rule.text,
      severity: rule.severity,
      passed: result.passed,
      message: result.message,
      evidence: result.evidence,
    };
    return { result, summary };
  }

  /**
   * 执行单个阶段的门禁检查（对齐 execute_stage）。
   *
   * 流程：检查 hard_rules（blocker 未通过 → blockerMessages，warn → warningMessages）
   * → 检查 pitfalls（可选阶段的 blocker 降级为 warning）→ 可选阶段的 hard_rule
   * blocker 也降级并清空 → 记录到 state.stageResults。
   */
  async executeStage(stageId: string, context: PredicateContext = {}): Promise<SOPStageResult> {
    const stage = this.getStage(stageId);
    if (!stage) {
      return {
        stageId,
        stageLabel: '',
        passed: false,
        hardRuleResults: [],
        pitfallResults: [],
        blockerMessages: [`stage '${stageId}' not found in SOP '${this.sop.id}'`],
        warningMessages: [],
        executedAt: new Date().toISOString(),
      };
    }

    const hardRuleResults: RuleSummary[] = [];
    const pitfallResults: RuleSummary[] = [];
    const blockerMessages: string[] = [];
    const warningMessages: string[] = [];

    // 检查 hard_rules
    for (const rule of stage.hardRules) {
      const { result, summary } = await this.checkRule(rule, context);
      hardRuleResults.push(summary);
      if (!result.passed) {
        const msg = `[${rule.id}] ${rule.text} — ${result.message}`;
        if (rule.severity === Severity.BLOCKER) {
          blockerMessages.push(msg);
        } else {
          warningMessages.push(msg);
        }
      }
    }

    // 检查 pitfalls
    for (const pitfall of stage.pitfalls) {
      const { result, summary } = await this.checkRule(pitfall, context);
      pitfallResults.push(summary);
      if (!result.passed) {
        const msg = `[${pitfall.id}] ${pitfall.text} — ${result.message}`;
        if (pitfall.severity === Severity.BLOCKER) {
          // pitfall 的 blocker 仅在非可选阶段才阻断
          if (stage.optional) {
            warningMessages.push(`(optional stage) ${msg}`);
          } else {
            blockerMessages.push(msg);
          }
        } else {
          warningMessages.push(msg);
        }
      }
    }

    // 可选阶段失败不阻断主流程
    let stagePassed = stage.optional ? true : blockerMessages.length === 0;
    if (stage.optional && blockerMessages.length > 0) {
      // 可选阶段的 blocker 降级为 warning
      for (const m of blockerMessages) {
        warningMessages.push(`(optional, downgraded) ${m}`);
      }
      blockerMessages.length = 0;
      stagePassed = true;
    }

    const stageResult: SOPStageResult = {
      stageId: stage.id,
      stageLabel: stage.label,
      passed: stagePassed,
      hardRuleResults,
      pitfallResults,
      blockerMessages,
      warningMessages,
      executedAt: new Date().toISOString(),
    };

    // 记录到执行状态
    this.state.stageResults[stage.id] = stageResult;
    return stageResult;
  }

  /**
   * 执行完整 SOP — 按顺序执行所有阶段（对齐 execute_sop）。
   *
   * 阶段未通过即 break（success=false）；全部通过则 success=true + completed。
   */
  async executeSop(featureId: string, context: PredicateContext = {}): Promise<SOPExecutionResult> {
    const startedAt = new Date().toISOString();
    this.state = {
      ...makeSOPExecutionState({ sopId: this.sop.id }),
      featureId,
      stageIndex: 0,
      startedAt,
      completed: false,
    };

    const stageResults: SOPStageResult[] = [];
    const allBlockerMessages: string[] = [];
    const allWarningMessages: string[] = [];
    let finalStageId = '';
    let success = false;

    let halted = false;
    for (let index = 0; index < this.sop.stages.length; index += 1) {
      const stage = this.sop.stages[index]!;
      this.state.stageIndex = index;
      finalStageId = stage.id;

      const stageResult = await this.executeStage(stage.id, context);
      stageResults.push(stageResult);
      allBlockerMessages.push(...stageResult.blockerMessages);
      allWarningMessages.push(...stageResult.warningMessages);

      if (!stageResult.passed) {
        halted = true;
        break;
      }
    }
    if (!halted) {
      // 所有阶段都通过（对齐 Python for-else：未 break）
      success = true;
      this.state.completed = true;
    }

    return {
      sopId: this.sop.id,
      featureId,
      success,
      stageResults,
      finalStageId,
      blockerMessages: allBlockerMessages,
      warningMessages: allWarningMessages,
      startedAt,
      completedAt: new Date().toISOString(),
    };
  }

  /** 重置执行状态到初始值（对齐 reset） */
  reset(): void {
    this.state = makeSOPExecutionState({ sopId: this.sop.id });
  }

  /** 从当前阶段继续执行（不重置 stageIndex，对齐 resume_from_current） */
  resumeFromCurrent(): void {
    this.state.completed = false;
  }
}

/* ------------------------------------------------------------------ */
/* YAML 加载（snake_case → camelCase，对齐 load_sop_from_yaml / dir） */
/* ------------------------------------------------------------------ */

function asRecord(value: unknown, what: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new Error(`SOP YAML invalid: ${what} must be a mapping`);
  }
  return value as Record<string, unknown>;
}

function asString(value: unknown, what: string): string {
  if (typeof value !== 'string') {
    throw new Error(`SOP YAML invalid: ${what} must be a string`);
  }
  return value;
}

function asStringArray(value: unknown, what: string): string[] {
  if (!Array.isArray(value) || value.some((v) => typeof v !== 'string')) {
    throw new Error(`SOP YAML invalid: ${what} must be a list of strings`);
  }
  return value as string[];
}

function parsePredicateConfig(raw: unknown, what: string): PredicateConfig {
  const data = asRecord(raw, what);
  const type = data.type;
  if (typeof type !== 'string' || type.length === 0) {
    throw new Error(`SOP YAML invalid: ${what}.type is required`);
  }
  return makePredicateConfig({
    type: type as PredicateConfig['type'],
    reason: data.reason !== undefined ? asString(data.reason, `${what}.reason`) : undefined,
    repository:
      data.repository !== undefined ? asString(data.repository, `${what}.repository`) : undefined,
    branch: data.branch !== undefined ? asString(data.branch, `${what}.branch`) : undefined,
    checks: data.checks !== undefined ? asStringArray(data.checks, `${what}.checks`) : undefined,
    beforeCommand:
      data.before_command !== undefined
        ? asString(data.before_command, `${what}.before_command`)
        : undefined,
    envVars:
      data.env_vars !== undefined ? asStringArray(data.env_vars, `${what}.env_vars`) : undefined,
    mustMatch:
      data.must_match !== undefined ? asString(data.must_match, `${what}.must_match`) : undefined,
    mustNotMatch:
      data.must_not_match !== undefined
        ? asString(data.must_not_match, `${what}.must_not_match`)
        : undefined,
    mustInclude:
      data.must_include !== undefined
        ? asStringArray(data.must_include, `${what}.must_include`)
        : undefined,
    antiPattern:
      data.anti_pattern !== undefined
        ? asStringArray(data.anti_pattern, `${what}.anti_pattern`)
        : undefined,
    cwdContains:
      data.cwd_contains !== undefined
        ? asString(data.cwd_contains, `${what}.cwd_contains`)
        : undefined,
    constraint:
      data.constraint !== undefined ? asString(data.constraint, `${what}.constraint`) : undefined,
  });
}

function parseRule(raw: unknown, what: string, defaultSeverity: Severity): HardRule | Pitfall {
  const data = asRecord(raw, what);
  if (typeof data.id !== 'string' || data.id.length === 0) {
    throw new Error(`SOP YAML invalid: ${what}.id is required`);
  }
  if (typeof data.text !== 'string') {
    throw new Error(`SOP YAML invalid: ${what}.text is required`);
  }
  const severity =
    data.severity !== undefined ? (asString(data.severity, `${what}.severity`) as Severity) : defaultSeverity;
  return {
    id: data.id,
    text: data.text,
    severity,
    predicate: parsePredicateConfig(data.predicate, `${what}.predicate`),
  };
}

function parseStage(raw: unknown, what: string): SOPStage {
  const data = asRecord(raw, what);
  if (typeof data.id !== 'string' || data.id.length === 0) {
    throw new Error(`SOP YAML invalid: ${what}.id is required`);
  }
  const hardRulesRaw = data.hard_rules;
  const pitfallsRaw = data.pitfalls;
  if (hardRulesRaw !== undefined && !Array.isArray(hardRulesRaw)) {
    throw new Error(`SOP YAML invalid: ${what}.hard_rules must be a list`);
  }
  if (pitfallsRaw !== undefined && !Array.isArray(pitfallsRaw)) {
    throw new Error(`SOP YAML invalid: ${what}.pitfalls must be a list`);
  }
  return makeSOPStage({
    id: data.id,
    label: data.label !== undefined ? asString(data.label, `${what}.label`) : undefined,
    suggestedSkill:
      data.suggested_skill !== undefined
        ? asString(data.suggested_skill, `${what}.suggested_skill`)
        : undefined,
    hardRules: hardRulesRaw?.map((r, i) => parseRule(r, `${what}.hard_rules[${i}]`, Severity.BLOCKER)),
    pitfalls: pitfallsRaw?.map((r, i) => parseRule(r, `${what}.pitfalls[${i}]`, Severity.WARN)),
    optional: data.optional !== undefined ? Boolean(data.optional) : undefined,
  });
}

/** 从解析后的 YAML 数据构造 SOPDefinition（对齐 SOPDefinition.model_validate） */
export function parseSopDefinition(raw: unknown): SOPDefinition {
  const data = asRecord(raw, 'SOP definition');
  if (typeof data.id !== 'string' || data.id.length === 0) {
    throw new Error('SOP YAML invalid: id is required');
  }
  const stagesRaw = data.stages;
  if (stagesRaw !== undefined && !Array.isArray(stagesRaw)) {
    throw new Error('SOP YAML invalid: stages must be a list');
  }
  return makeSOPDefinition({
    id: data.id,
    domain: data.domain !== undefined ? asString(data.domain, 'domain') : undefined,
    label: data.label !== undefined ? asString(data.label, 'label') : undefined,
    description:
      data.description !== undefined ? asString(data.description, 'description') : undefined,
    stages: stagesRaw?.map((s, i) => parseStage(s, `stages[${i}]`)),
  });
}

/** 从 YAML 文件加载 SOP 定义（对齐 load_sop_from_yaml） */
export async function loadSopFromYaml(yamlPath: string): Promise<SOPDefinition> {
  const text = await fs.readFile(resolve(yamlPath), 'utf-8');
  return parseSopDefinition(parseYaml(text));
}

/**
 * 从目录加载所有 SOP 定义（对齐 load_sops_from_dir）。
 *
 * 仅扫描 *.yaml（与 Python glob 一致），单个文件解析失败跳过不中断。
 */
export async function loadSopsFromDir(sopsDir: string): Promise<Map<string, SOPDefinition>> {
  const sops = new Map<string, SOPDefinition>();
  let entries: string[];
  try {
    entries = await fs.readdir(resolve(sopsDir));
  } catch {
    return sops;
  }
  for (const name of entries.sort()) {
    if (!name.endsWith('.yaml')) {
      continue;
    }
    try {
      const sop = await loadSopFromYaml(resolve(sopsDir, name));
      sops.set(sop.id, sop);
    } catch {
      // 对齐 Python：单文件失败仅记录，不中断目录加载
    }
  }
  return sops;
}

/** 目录内 yaml 文件名（不含扩展名），便于日志/诊断 */
export function sopFileName(yamlPath: string): string {
  return basename(yamlPath, '.yaml');
}
